### PR TITLE
feat(ludicrous): Run mutations from the same predicate concurrently in ludicrous mode

### PR DESCRIPTION
**Pull request by harshil-goel**
Added topological sort to ludicrous mode executor to parallelize execution in one predicate. 
Master: 21 Million dataset: 3m 15 sec
This PR: 21 Million dataset: 2m 25 sec

Fixes #DGRAPH-1357

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6060)

 
 
 

Docs Preview: [Dgraph Preview](https://dgraph-2093d2b5e7-89327.surge.sh)

### CODE REVIEW

**Review by harshil-goel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6060)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @manishrjain)

___
*[worker/executor.go, line 71 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MDLiH7U49-uel5-N9LM:-MFZIFgi1JWLiVOn-p9z:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7eb3f3c64a9a6f8a7d3b9246e0c02b7669eed558/worker/executor.go#L71)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Move this closer to usage.

var keys []uint64
</blockquote></details>

Done.
___
*[worker/executor.go, line 81 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MDLjcZD9DTtu4El53zf:-MFZIGq_AksZc5_DuOTz:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7eb3f3c64a9a6f8a7d3b9246e0c02b7669eed558/worker/executor.go#L81)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

If you have count index, only then do serially. Not on IsList.

use `uniq[0]`
</blockquote></details>

Done.
___
*[worker/executor.go, line 180 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MDLjC724P5us-IPo84a:-MFZIHqOBUU6bBajf7Gu:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7eb3f3c64a9a6f8a7d3b9246e0c02b7669eed558/worker/executor.go#L180)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Maybe see if you don't need channels. You could just create goroutines to do the work. Do use throttle to limit how many goroutines you do create.
</blockquote></details>

Done.
___
*[worker/executor.go, line 193 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MDLjJl31JNmgCn8bYZp:-MFZIJ420bDUr1qYlSnY:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7eb3f3c64a9a6f8a7d3b9246e0c02b7669eed558/worker/executor.go#L193)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

100 chars.
</blockquote></details>

Done.

**Review by ashish-goswami — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6060)* status: 0 of 2 files reviewed, 19 unresolved discussions (waiting on @harshil-goel, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[posting/list.go, line 437 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFebCKV8ia5OdN9XSyb:-MFebCKV8ia5OdN9XSyc:b-plwlcw) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/posting/list.go#L437)):*
> ```Go
> }
> 
> func GetConflictKeys(pk x.ParsedKey, key []byte, t *pb.DirectedEdge) uint64 {
> ```

Its returning single uint64, should we call it `GetConflictKey` instead of `GetConflictKeys`?
___
*[worker/executor.go, line 59 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFePcDw5fZv5Ab9MJR4:-MFePcDw5fZv5Ab9MJR5:b-3fnx69) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L59)):*
> ```Go
> 
> func newExecutor(applied *y.WaterMark) *executor {
> 	runtime.SetBlockProfileRate(1)
> ```

Remove this line.
___
*[worker/executor.go, line 64 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFebPD66gFhx7EdI6SS:-MFebPD66gFhx7EdI6ST:b-9qiiat) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L64)):*
> ```Go
> 		closer:   y.NewCloser(0),
> 		applied:  applied,
> 		throttle: y.NewThrottle(2000),
> ```

Comment about default value of `2000` would be nice.
___
*[worker/executor.go, line 72 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFe_lKoDrshLBFmcqV-:-MFe_lKoDrshLBFmcqV0:bnhfjol) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L72)):*
> ```Go
> 
> func generateTokenKeys(nq *pb.DirectedEdge, tokenizers []tok.Tokenizer) ([]uint64, error) {
> 	keys := make([]uint64, len(tokenizers))
> ```

Looks it it should be `keys := make([]uint64, 0, len(tokenizers))`
___
*[worker/executor.go, line 82 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFe_y4i5K8yd9Lgeq2a:-MFe_y4i5K8yd9Lgeq2b:b-zhw1bi) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L82)):*
> ```Go
> 		schemaVal, err := types.Convert(storageVal, types.TypeID(nq.GetValueType()))
> 		if err != nil {
> 			errs = append(errs, err.Error())
> ```

instead of calling `err.Error()` lets have []error only at top.
Also if there is an error here, should be continue further processing?
___
*[worker/executor.go, line 87 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFeaVjyBXDDTeBeQQiS:-MFeaVjz-czs7smWDvLa:bu3ys0o) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L87)):*
> ```Go
> 			nq.Lang))
> 		if err != nil {
> 			errs = append(errs, err.Error())
> ```

same here should we continue processing here even if there is error?
___
*[worker/executor.go, line 122 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFecqt06M67Qn7GTuut:-MFecqt06M67Qn7GTuuu:b-t3t969) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L122)):*
> ```Go
> 
> 		keys, err := generateTokenKeys(edge, tokenizers)
> 		for _, i := range keys {
> ```

instead of i call it key
___
*[worker/executor.go, line 138 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf8diHB6gnFJ5dZWh3:-MFf8diHB6gnFJ5dZWh4:b1hdo78) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L138)):*
> ```Go
> 
> type mutation struct {
> 	m     *subMutation
> ```

we can call it sm.
___
*[worker/executor.go, line 139 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf6cw32uSQLTKM0ani:-MFf6cw32uSQLTKM0anj:b-qz6seg) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L139)):*
> ```Go
> type mutation struct {
> 	m     *subMutation
> 	keys  []uint64
> ```

we can call it conflictKeys.
___
*[worker/executor.go, line 142 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf6Y4k46ltMp2ebksl:-MFf6Y4k46ltMp2ebksm:bn5xevs) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L142)):*
> ```Go
> 	inDeg int
> 
> 	outEdges map[uint64]*mutation
> ```

we should call it dependent mutations.
___
*[worker/executor.go, line 194 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFeeQ3XClbDc5bqgnPz:-MFeeQ3XClbDc5bqgnQ-:bbqzu6a) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L194)):*
> ```Go
> 		dependent.inDeg -= 1
> 		if dependent.inDeg == 0 {
> 			x.Check(e.throttle.Do())
> ```

Not sure, if we should crash here.
___
*[worker/executor.go, line 197 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf754Y4pnfx3CUQi9m:-MFf754Y4pnfx3CUQi9n:b-bs8ky6) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L197)):*
> ```Go
> 			go func(d *mutation) {
> 				e.worker(d)
> 			}(dependent)
> ```

`go e.worker(dependent)`
___
*[worker/executor.go, line 202 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf8TFCAxOFqdFdwtZO:-MFf8TFCAxOFqdFdwtZP:bqzlflp) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L202)):*
> ```Go
> 
> 	for _, c := range mut.keys {
> 		i := 0
> ```

we should rename it to something else.
___
*[worker/executor.go, line 205 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf8NS-8t3PFfLBo2rE:-MFf8NS-8t3PFfLBo2rF:bxvfuth) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L205)):*
> ```Go
> 		arr := mut.graph.conflicts[c]
> 
> 		for _, x := range arr {
> ```

Add a comment that we are deleting the mutation here.
___
*[worker/executor.go, line 261 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf68LJ83RsAGzUI2EM:-MFf68LJ83RsAGzUI2EN:b-lamluo) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L261)):*
> ```Go
> 			x.Check(e.throttle.Do())
> 			go func() {
> 				e.worker(m)
> ```

`go e.worker(m)`

**Review by harshil-goel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6060)* status: 0 of 3 files reviewed, 19 unresolved discussions (waiting on @ashish-goswami, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[posting/list.go, line 437 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFebCKV8ia5OdN9XSyb:-MFkZUJs9WBPAv5JuVch:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/posting/list.go#L437)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

Its returning single uint64, should we call it `GetConflictKey` instead of `GetConflictKeys`?
</blockquote></details>

Done.
___
*[worker/executor.go, line 59 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFePcDw5fZv5Ab9MJR4:-MFkZV2I6scL3gPDPZv0:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L59)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

Remove this line.
</blockquote></details>

Done.
___
*[worker/executor.go, line 64 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFebPD66gFhx7EdI6SS:-MFkZVq-6ELatIZKGJf5:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L64)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

Comment about default value of `2000` would be nice.
</blockquote></details>

Done.
___
*[worker/executor.go, line 72 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFe_lKoDrshLBFmcqV-:-MFkZc6pDvqREj4xMf_T:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L72)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

Looks it it should be `keys := make([]uint64, 0, len(tokenizers))`
</blockquote></details>

Done.
___
*[worker/executor.go, line 82 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFe_y4i5K8yd9Lgeq2a:-MFkZcigB2zsHnD0uvHc:bdvniai) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L82)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

instead of calling `err.Error()` lets have []error only at top.
Also if there is an error here, should be continue further processing?
</blockquote></details>

Yeah,  We just care about whatever conflict keys we can generate, to best avoid conflicting mutations running together. 
I used `error` instead of `string` because we need to join these errors later, and would need to convert to string to do that. 
___
*[worker/executor.go, line 87 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFeaVjyBXDDTeBeQQiS:-MFk_KcgCIk-cCFse59O:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L87)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

same here should we continue processing here even if there is error?
</blockquote></details>

Done.
___
*[worker/executor.go, line 122 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFecqt06M67Qn7GTuut:-MFk_LZ71yy6grYR-PYA:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L122)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

instead of i call it key
</blockquote></details>

Done.
___
*[worker/executor.go, line 138 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf8diHB6gnFJ5dZWh3:-MFk_NTz9DQuc_LihXNy:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L138)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

we can call it sm.
</blockquote></details>

Done.
___
*[worker/executor.go, line 139 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf6cw32uSQLTKM0ani:-MFk_Nmj7GFcEfd6_1vZ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L139)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

we can call it conflictKeys.
</blockquote></details>

Done.
___
*[worker/executor.go, line 142 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf6Y4k46ltMp2ebksl:-MFk_Xk1Ajee6wJ9IDxd:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L142)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

we should call it dependent mutations.
</blockquote></details>

Done.
___
*[worker/executor.go, line 194 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFeeQ3XClbDc5bqgnPz:-MFk_TGJAA8nst4_RVvJ:bp6wr0s) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L194)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

Not sure, if we should crash here.
</blockquote></details>

We always return nil, so it wouldn't crash. If it crashes, it's a bug. 
___
*[worker/executor.go, line 197 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf754Y4pnfx3CUQi9m:-MFk_SqWEIZ4hh9rH9HK:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L197)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

`go e.worker(dependent)`
</blockquote></details>

Done.
___
*[worker/executor.go, line 202 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf8TFCAxOFqdFdwtZO:-MFk_SNFDadWS_l0Yt-a:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L202)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

we should rename it to something else.
</blockquote></details>

Done.
___
*[worker/executor.go, line 205 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf8NS-8t3PFfLBo2rE:-MFk_RuBAQtrD4dz9Rbb:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L205)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

Add a comment that we are deleting the mutation here.
</blockquote></details>

Done.
___
*[worker/executor.go, line 261 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MFf68LJ83RsAGzUI2EM:-MFk_Qi89HCRl2RC5Q4F:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fd5fee77b16618df31222d27031998d1678e9223/worker/executor.go#L261)):*
<details><summary><i>Previously, ashish-goswami (Ashish Goswami) wrote…</i></summary><blockquote>

`go e.worker(m)`
</blockquote></details>

Done.

File: worker/executor.go [hunk 1]
```diff
@@ -44,49 +51,208 @@ type executor struct {
...
 	applied  *y.WaterMark
+	throttle *y.Throttle
 }
 
 func newExecutor(applied *y.WaterMark) *executor {
 	ex := &executor{
 		predChan: make(map[string]chan *subMutation),
 		closer:   y.NewCloser(0),
 		applied:  applied,
+		throttle: y.NewThrottle(2000), // Run 2000 mutations at a time.
```
Comment (parasssh) [inline worker/executor.go:62]: should we make this configurable?

File: worker/executor.go [hunk 2]
```diff
@@ -44,49 +51,208 @@ type executor struct {
...
+		keys, err := generateTokenKeys(edge, tokenizers)
+		for _, key := range keys {
+			uniq[key] = struct{}{}
+		}
+		if err != nil {
+			glog.V(2).Info("Error in generating tokens", err)
+		}
+	}
+
+	keys := make([]uint64, 0)
```
Comment (parasssh) [inline worker/executor.go:127]: as discussed, we can just return the map instead of the slice.

**Review by harshil-goel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6060)* status: 0 of 5 files reviewed, 21 unresolved discussions (waiting on @ashish-goswami, @manishrjain, @martinmr, @parasssh, and @vvbalaji-dgraph)

___
*[worker/executor.go, line 62 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MCvya0L8icbwM2mnhj0-r5-62:-MG36YQe54USk6UJPXV8:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/686f103ffc4ddd175943b305afd3b89242c7c69b/worker/executor.go#L62)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

should we make this configurable?
</blockquote></details>

Done.
___
*[worker/executor.go, line 127 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MCvya0L8icbwM2mnhj0-r5-127:-MG36_2e9BLmRN8JgqaA:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/686f103ffc4ddd175943b305afd3b89242c7c69b/worker/executor.go#L127)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

as discussed, we can just return the map instead of the slice. 
</blockquote></details>

Done.

**Review by jarifibrahim — COMMENTED** [review]
Got some comments.

> Reviewed 1 of 2 files at r5, 1 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-:-MG3MVvW5zh_divETk60:b17ku4n)* status: 2 of 5 files reviewed, 25 unresolved discussions (waiting on @ashish-goswami, @harshil-goel, @manishrjain, @martinmr, @parasssh, and @vvbalaji-dgraph)

___
*[worker/executor.go, line 94 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MG3M0z7EfeSTsNxyJyj:-MG3M0z8BooNgt0gfHgB:brbwgak) ([raw file](https://github.com/dgraph-io/dgraph/blob/e038ccd1dcc8c2b547ff8e062a41260bc05e4701/worker/executor.go#L94)):*
> ```Go
> 
> 	if len(errs) > 0 {
> 		return keys, fmt.Errorf(strings.Join(errs, "\n"))
> ```

Shouldn't his be `nil,  fmt....`
___
*[worker/executor.go, line 106 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MG3M5ZRDcxSq_aA3K2c:-MG3M5ZRDcxSq_aA3K2d:bxhtz1q) ([raw file](https://github.com/dgraph-io/dgraph/blob/e038ccd1dcc8c2b547ff8e062a41260bc05e4701/worker/executor.go#L106)):*
> ```Go
> 		pk, err := x.Parse(key)
> 		if err != nil {
> 			continue
> ```

Why are we skipping an error? Do we expect x.Parse to fail in any case?
___
*[worker/executor.go, line 118 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MG3Ldg1D0EtO9DkPKOJ:-MG3Ldg1D0EtO9DkPKOK:bjn7zrq) ([raw file](https://github.com/dgraph-io/dgraph/blob/e038ccd1dcc8c2b547ff8e062a41260bc05e4701/worker/executor.go#L118)):*
> ```Go
> 		}
> 
> 		tokens, err := generateTokenKeys(edge, tokenizers)
> ```

err should be checked before you use `tokens`.
___
*[worker/executor.go, line 181 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MG394H12AFdYREW83Z6:-MG394H12AFdYREW83Z7:b-tr4wh0) ([raw file](https://github.com/dgraph-io/dgraph/blob/e038ccd1dcc8c2b547ff8e062a41260bc05e4701/worker/executor.go#L181)):*
> ```Go
> 	e.throttle.Done(nil)
> 
> 	mut.graph.Lock()
> ```

defer mut.graph.unlock(). Right now the unlock is at the end of the function which makes it hard to find.

**Review by harshil-goel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6060)* status: 2 of 5 files reviewed, 25 unresolved discussions (waiting on @ashish-goswami, @jarifibrahim, @manishrjain, @martinmr, @parasssh, and @vvbalaji-dgraph)

___
*[worker/executor.go, line 94 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MG3M0z7EfeSTsNxyJyj:-MG3aV63BbLfc2XqYHIT:bwu5zfq) ([raw file](https://github.com/dgraph-io/dgraph/blob/e038ccd1dcc8c2b547ff8e062a41260bc05e4701/worker/executor.go#L94)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Shouldn't his be `nil,  fmt....`
</blockquote></details>

No. We need all the keys we can get to make sure the data's safe.
___
*[worker/executor.go, line 106 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MG3M5ZRDcxSq_aA3K2c:-MG3aZvF-Rqu0bogaag1:b-sx86i5) ([raw file](https://github.com/dgraph-io/dgraph/blob/e038ccd1dcc8c2b547ff8e062a41260bc05e4701/worker/executor.go#L106)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Why are we skipping an error? Do we expect x.Parse to fail in any case?
</blockquote></details>

No, we don't. But let's log in any case. 
___
*[worker/executor.go, line 118 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MG3Ldg1D0EtO9DkPKOJ:-MG3ad706TPKQJz_SArC:bzb6x5x) ([raw file](https://github.com/dgraph-io/dgraph/blob/e038ccd1dcc8c2b547ff8e062a41260bc05e4701/worker/executor.go#L118)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

err should be checked before you use `tokens`.
</blockquote></details>

We continue if the err is not nil, err at this point is always nil. 
___
*[worker/executor.go, line 181 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MG394H12AFdYREW83Z6:-MG3am_8DYDL2fVKJ8mR:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/e038ccd1dcc8c2b547ff8e062a41260bc05e4701/worker/executor.go#L181)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

defer mut.graph.unlock(). Right now the unlock is at the end of the function which makes it hard to find.
</blockquote></details>

Done.

**Review by jarifibrahim — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-:-MG7xlHe50aC0G8fWlwy:bnfp4nl)* status: 2 of 5 files reviewed, 23 unresolved discussions (waiting on @ashish-goswami, @harshil-goel, @jarifibrahim, @manishrjain, @martinmr, @parasssh, and @vvbalaji-dgraph)

___
*[worker/executor.go, line 106 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/6060#-MG7xRWZ184_S94QC5RA:-MG7xRW_DIfOF-Jjg3eo:b-7uzieq) ([raw file](https://github.com/dgraph-io/dgraph/blob/b3400c811902bb22bc16fe3f5b8c9e1f1b2d70aa/worker/executor.go#L106)):*
> ```Go
> 		pk, err := x.Parse(key)
> 		if err != nil {
> 			glog.V(2).Info("Error in generating conflic keys", err)
> ```

conflic => conflict

**Pull request merged**